src/storage/src/upsert.rs

```diff
@@ -98,15 +98,9 @@ impl UpsertKey {
 
     pub fn from_value(value: Result<&Row, &UpsertError>, mut key_indices: &[usize]) -> Self {
         Self::from_iter(value.map(|value| {
-            value.iter().enumerate().flat_map(move |(idx, datum)| {
-                let key_idx = key_indices.get(0)?;
-                if idx == *key_idx {
-                    key_indices = &key_indices[1..];
-                    Some(datum)
-                } else {
-                    None
-                }
-            })
+            key_indices
+                .iter()
+                .map(|&idx| value.iter().nth(idx).expect("must exist"))
```

Comment:
I'd recommend using a `DatumVec` here to avoid repeatedly decoding the row. Otherwise LGTM.

Reply:
I thought about it and then got lazy. I'll add that